sotodlib/io/bookbinder.py

```diff
@@ -563,6 +481,91 @@ def replace_times_and_trim_frame(self, f):
+            analysis. Set return_flags = True to return.
+        """
+        # There is no _prev_smurf_sample at the beginning of the book
+        if self._prev_smurf_sample is None:
+            prev_frame_end = self.BOOK_START_TIME.time - 1
+        else:
+            prev_frame_end = self._prev_smurf_sample.time
+        # Define a strict upper limit (exclusive) up until which to look/interpolate for missing samples
+        if self.BOOK_END_TIME is not None:
+            end_time = min(flush_time.time, self.BOOK_END_TIME.time)
+        elif self._smurf_timestamps is not None:
+            end_time = min(flush_time.time, self._smurf_timestamps[-1] + 1)
+        else:
+            end_time = min(flush_time.time, data.times[-1].time + 1)
+
+        # Obtain list of filled-in timestamps, whether from provided list or estimate with fill_time_gaps
+        if self._smurf_timestamps is not None:
+            # trim the list of timestamps to the current frame
+            m_ = np.logical_and(self._smurf_timestamps < end_time, self._smurf_timestamps > prev_frame_end)
+            ts = self._smurf_timestamps[m_]
+        else:
+            # fallback when no timestamp is provided. We switch to estimate sample interval based on data
+            assert data.times[0].time >= prev_frame_end
+            # Use fill_time_gaps to fill in any missing samples in the middle, as well as
+            # at the beginning (between previous frame and current frame)
+            dt = np.median(np.diff(data.times))  # estimate sample interval
+            if self._prev_smurf_sample is None:
+                # Use estimated sample interval to fill backward to prev_frame_end = book start time - 1
+                # since book start time does not usually line up with sample interval
+                ts = np.append(np.flip(np.arange(data.times[0].time-dt, prev_frame_end, -dt, dtype=int)), data.times)
+                ts = fill_time_gaps(ts)
+            elif (data.times[0].time - prev_frame_end)/dt > 0.5:
+                ts = fill_time_gaps(np.insert(data.times, 0, prev_frame_end))[1:]
+            else:
+                ts = fill_time_gaps(np.asarray(data.times))
+            # Use estimated sample interval to fill until end_time. Not using fill_time_gaps because
+            # end_time is externally determined and does not usually line up with sample interval
+            ts = np.append(ts, np.arange(data.times[-1].time+dt, end_time, dt, dtype=int))
+
+        # Create new G3SuperTimestream with filled-in samples
+        if len(data.times) < len(ts):
+            dt = np.median(np.diff(ts))
+            i_missing, _ = find_missing_samples(ts, np.array(data.times), atol=dt/2)
+            m = np.ones(len(ts), dtype=bool)
+            m[i_missing] = False
+            assert np.sum(m) == len(data.times)
+            new_data = np.full((data.data.shape[0], len(ts)), self.FLAGGED_SAMPLE_VALUE, dtype=data.data.dtype)
```

Comment:
I thought we wanted to interpolate here in the case of dropped samples? Like Katie said, setting to a FLAGGED_SAMPLE_VALUE isn't really helpful since that value can be valid detector data, so it can't be used to create a gap mask.

Reply:
I’m not following why interpolated values would make it easier to create a gap mask—and isn’t that what the flag array is for?

Reply:
People will need to interpolate through the gaps to use the data anyway, so writing interpolated values to disk (with the flag mask that tells you which values were interpolated) will reduce the preprocessing people will need to do on load.

Reply:
I’m missing something here—why do people need to interpolate through the gap to use it? For example, if I have:

```
Data:   1   2  -1   4   5
Flag:   0   0   1   0   0
```

why would having a -1 (or any other value) instead of 3 cause problems when the flag is there to say “ignore this value”?

Reply:
If we don't put the interpolation here it's going in the loader b/c this isn't the type of operation we have folks do themselves, it's one that just gets automated somewhere. I'd rather we do it here so it only has to happen once instead of on every load. 

We do need to document the data/flags setup we end up having in the tod format doc, because it will be important to be able to trace everything through.

Reply:
If you want to do something like take an FFT of your timestream, it's very important that both there aren't random -1's in the data, and that the data is equisampled (so you can't just mask out the -1's). Gap-filling + interpolation is generally how we handle that

Reply:
yes, if you want to do things like run a Fourier filter on the timestreams it's much easier to just have "mostly right" data points in there. Then you can run the filter without much concern able contaminating adject points. 

Otherwise you have to run Fourier filters accounting for un-equal sample spacing which is a HUGE efficiency hit.

Reply:
I see. In that case, I suppose we can just copy what Jack already wrote in `fill_time_gaps` (the last block). I won’t have time to do this though. (And the unit tests will also need to be updated accordingly.)

Reply:
lol Jack beat me to it (didn't see that it had been updated before my comment)